devicedefender/include/aws/iotdevicedefender/DeviceDefender.h

```diff
@@ -98,6 +173,57 @@ namespace Aws
                 void *cancellationUserdata = nullptr) noexcept;
 
             static void s_onDefenderV1TaskCancelled(void *userData);
+
+            struct customMetricData
+            {
+                size_t index;
+                ReportTask *task;
+            };
+            static int s_getCustomMetricNumber(double *output, void *customData);
```

Comment:
Overall, this approach feels very C-like and I think it could be made a lot cleaner using C++ semantics.

My suggestion would be to:

1. Make a simple, flat class hierarchy: CustomMetricBase -> {CustomMetricNumber, CustomMetricNumberList, CustomMetricStringList, CustomMetricIpList}
2. Each leaf would hold a moved version of the passed in gathering function as well as a dynamically constructed function that we'll pass in to aws-c-iot that has captured the ReportTask context via lambda capture so there's no need for indexing or backpointers or anything like that.  See std::function::target for converting between std::function and C function pointer.
3. Have a single vector of CustomMetricBase *